File: 20230615_Thursday/20230615.rb
# Leetcode: 1161. Maximum Level Sum of a Binary Tree.
# https://leetcode.com/problems/maximum-level-sum-of-a-binary-tree/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 279 ms, faster than 7.14% of Ruby online submissions for Maximum
# Level Sum of a Binary Tree.
# Memory Usage: 214.6 MB, less than 100.00% of Ruby online submissions for
# Maximum Level Sum of a Binary Tree.
# 2023.06.16 Updated.
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer}
def max_level_sum(root)
  q = [[root,1]]
  max = root.val
  l = 1
  while !q.empty?
    sum = 0
    nq = []
    current_level = q[0].last 
    q.each do |node,lvl|
      sum += node.val 
      nq.push([node.left,lvl+1]) if node.left
      nq.push([node.right,lvl+1]) if node.right
    end
    q = nq
    if max < sum 
      max = sum
      l = current_level
    end
  end
  l
end

File: 20230615_Thursday/20230615.go
// Leetcode: 1161. Maximum Level Sum of a Binary Tree.
// https://leetcode.com/problems/maximum-level-sum-of-a-binary-tree/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 112 ms, faster than 79.80% of Go online submissions for Maximum
// Level Sum of a Binary Tree.
// Memory Usage: 7.9 MB, less than 62.63% of Go online submissions for Maximum
// Level Sum of a Binary Tree.
// 2023.06.16 Updated.
package main

import (
	"fmt"
)

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxLevelSum(root *TreeNode) int {
	q := make([]*TreeNode, 1)
	current_level := 0
	q[0] = root
	max := root.Val
	ans := 1
	for 0 != len(q) {
		var nq []*TreeNode
		current_level += 1
		sum := 0
		for _, node := range q {
			sum += node.Val
			if nil != node.Left {
				nq = append(nq, node.Left)
			}
			if nil != node.Right {
				nq = append(nq, node.Right)
			}
		}
		q = nq
		if max < sum {
			max = sum
			ans = current_level
		}
	}
	return ans
}
